File: 0159. Longest Substring with At Most Two Distinct Characters.java
// Medium
// Two Pointers, Hash Table
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/longest-substring-with-at-most-two-distinct-characters/

import java.util.*;
class Solution {
  // Use HashMap to count character frequencies in the window
  // Expand right pointer to include new characters
  // If there are more than two distinct chars, move left to shrink the window
  public int lengthOfLongestSubstringTwoDistinct(String s) {
      Map<Character, Integer> map = new HashMap<>();
      int left = 0, maxLen = 0;
      for (int right = 0; right < s.length(); right++) {
          char c = s.charAt(right);
          map.put(c, map.getOrDefault(c, 0) + 1);
          while (map.size() > 2) {
              char leftChar = s.charAt(left);
              map.put(leftChar, map.getOrDefault(leftChar, 0) - 1);
              if (map.get(leftChar) == 0) {
                  map.remove(leftChar);
              }
              left++;
          }
          maxLen = Math.max(maxLen, right - left + 1);
      }
      return maxLen;
  }
}

File: 0001-0500/0159. Longest Substring with At Most Two Distinct Characters.py
# Time:O(n), Space:O(1)

class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        if len(s) < 3:
            return len(s)
        
        left = 0
        max_len = 0
        char_count = {}

        for right in range(len(s)):
            char_count[s[right]] = char_count.get(s[right], 0) + 1

            while len(char_count) > 2:
                char_count[s[left]] -= 1
                if char_count[s[left]] == 0:
                    del char_count[s[left]]
                left += 1
            max_len = max(max_len, right - left + 1)
            
        return max_len
    